### PR TITLE
Threaded calls to ElasticSearch

### PR DESCRIPTION
**Pull request by sriamazingram**
Modified candidate_generation/utility.py to make candidate generation calls in a threaded manner.

### CODE REVIEW

File: tl/candidate_generation/utility.py [hunk 1]
```diff
@@ -36,17 +37,27 @@ def create_candidates_df(self, df, column, size, properties, method, lower_case=
...
                                 c not in ['kg_id', 'kg_labels', 'method', 'kg_descriptions',
                                           self.previous_match_column_name]]
+            tuples = list()
             for key_tuple, gdf in grouped:
                 gdf.reset_index(inplace=True)
-                tuple = ((c, gdf.at[0, c]) for c in relevant_columns)
-
+                tuples.append(((c, gdf.at[0, c]) for c in relevant_columns))
+            
+            def create_cf_candidates_threaded(tuple):
```
Comment (saggu) [inline tl/candidate_generation/utility.py:45]: create this method as a class method and not as a nested function.

File: tl/candidate_generation/utility.py [hunk 2]
```diff
@@ -60,15 +71,16 @@ def create_cfd_canonical(self, df, relevant_columns, column, size, properties, m
         candidates_format = list()
         all_candidates_aux_dict = {}
 
-        for i, row in df.iterrows():
+        def call_es(row):
```
Comment (saggu) [inline tl/candidate_generation/utility.py:74]: same for this method. The code needs to be restructured so that only one threaded function is implemented and called in both canonical and candidates method

File: tl/candidate_generation/utility.py [hunk 3]
```diff
@@ -100,6 +112,13 @@ def create_cfd_canonical(self, df, relevant_columns, column, size, properties, m
                     cf_dict['pagerank'] = candidate_dict[kg_id]['pagerank_float']
                     cf_dict[self.score_column_name] = candidate_dict[kg_id]['score']
                     candidates_format.append(cf_dict)
+        
+        max_threads = 50
```
Comment (saggu) [inline tl/candidate_generation/utility.py:116]: This should be:
1. A parameter with a default value of 50
2. min(number of rows in the file, 50)

**Review by saggu — CHANGES_REQUESTED** [review]
Also, make sure the code is PEP 8 compatible

File: tl/candidate_generation/utility.py [hunk 4]
```diff
@@ -168,5 +151,5 @@ def write_auxiliary_files(self, auxiliary_folder, all_candidates_aux_dict, auxil
 
             for key in _:
                 df = pd.DataFrame(_[key])
-                if len(df) > 0:
```
Comment (saggu) [inline tl/candidate_generation/utility.py:171]: why did you remove this check?

**Review by saggu — CHANGES_REQUESTED** [review]
Is the code properly tested? There are a lots of changes. 
1. Pick a couple of files. 
2. Run only one candidate generation command on each of the files
3. Run both candidate generation commands in sequence on each of the files
Let's discuss the output of the above experiment.

File: tl/candidate_generation/utility.py (hunk 4 above)
Comment (sriamazingram) [reply tl/candidate_generation/utility.py:171]: Actually, this check was not there in the master branch, which I had cloned for my experiments. So it did not show up in my fork. Should I add it to the function?

Comment (saggu) [reply tl/candidate_generation/utility.py:171]: you should sync your fork with `development` first and then create this pull request. I will close this PR

**Comment by saggu** [comment]
@sriamazingram needs to sync his fork with development
